### PR TITLE
Refactoring Edit Flow

### LINKED ISSUE #1007: [Overlapping] Creating an overlapping strategy shows 'Recurring' strategy type in strategy creation page
1. Create a new strategy
2. Choose overlapping type
4. In creation [page](https://app.carbondefi.xyz/strategies/create?base=0x1F573D6Fb3F13d689FF844B4cE37794d79a7FF1C&quote=0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE&strategyType=recurring&strategySettings=overlapping), the type under the token pair shows as 'Recurring'

### LINKED ISSUE #1204: [Refactoring] Split Creation / Edit flow into its own pages
Currently the edit & creation flow are only 2 pages. It makes it very difficult to create custom design & logic for each form.

## Goals 
- Simplify behavior specific to each strategy
- Reduce technical debt
- Avoid side effect when adding a new feature
- Keep state on reload
- Share prefilled strategy

## Tasks
- [ ] Token selection should be its own page

- [ ] Split create / deposit / withdraw & edit price into 4 pages per type of strategy :
- `create/overlapping`
- `edit/deposit/overlapping`
- `edit/withdraw/overlapping`
- `edit/price/overlapping`
-> Same for `recurring` and `disposable`

- [ ] Use the search params URL to keep state of the fields

For overlapping :
- marketPrice
- min
- max
- spread
- budget
- anchor

For recurring :
- buy.min
- buy.max
- buy.budget
- sell.min
- sell.max
- sell.budget
-> We can derive limit/range from min/max

For disposable: same are recurring with 
- direction

Extract logic from `useCreateStrategy`: 
- #1223

### LINKED ISSUE #1223: [Overlapping] Extract spread out of useCreateStrategy
Currently we use spread in the `useCreateStrategy` because of the duplicate method.
It would make more sense to move it out of it and only use it in the create overlapping component

### LINKED ISSUE #1248: [Refactoring] Split Edit flow into its own pages
Split Edit flow into multiple pages:
- `strategies/edit/:strategyId/recurring/renew`
- `strategies/edit/:strategyId/recurring/prices`
- `strategies/edit/:strategyId/recurring/deposit`
- `strategies/edit/:strategyId/recurring/withdraw`
- `strategies/edit/:strategyId/overlapping/renew`
- `strategies/edit/:strategyId/overlapping/prices`
- `strategies/edit/:strategyId/overlapping/deposit`
- `strategies/edit/:strategyId/overlapping/withdraw`

### LINKED ISSUE #1257: Create strategy - overlapping strategy has a "recurring" tag
flow:
create a new overlapping strategy
on the page where you set prices, it says "recurring".
image

updated:
replace it with "overlapping"

### PR DESCRIPTION
**Pull request by GrandSchtroumpf**
fix #1204
fix #1201
fix #1223

fix #1248
fix #1257
fix #1007

close #1201 

- [x] Split creation flow into 4 pages (token selection, disposable, recurring, overlapping)
- [x] Split Edit Price into 3 pages: disposable, recurring, overlapping
- [x] Split Edit Budget into 3 pages: disposable, recurring, overlapping

### CODE REVIEW

**Comment by ashachaf** [comment]
- [ ] price reset flow
https://github.com/bancorprotocol/carbon-app/assets/40430285/4c4166c2-8f4a-4cba-80e6-c0b4570e8841

- [ ] overlapping chart is "slow" to show 
https://github.com/bancorprotocol/carbon-app/assets/40430285/a9900f53-c4df-4bfc-bdf2-6f433c1f36a3

- [ ] deposit into overlapping strategy error
flow:
try to deposit into an overlapping strategy and after clicking the cta, i have console error and stay on the same page.
image

- [ ] deposit into overlapping strategy - no budget calculation indication 
expected: when updating prices of an overlapping strategy, we calculate the 2nd budget on behalf of the user and indicate in the progress bar visual. 
actual:
https://github.com/bancorprotocol/carbon-app/assets/40430285/0107b2da-33e1-4b0e-9961-69c03cbf885d

File: src/components/strategies/create/CreateForm.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  order1: BaseOrder;
+  approvalText?: string;
+  children: ReactNode;
+}
+
+export const CreateForm: FC<FormProps> = (props) => {
+  const { base, quote, order0, order1, type, children } = props;
+  const { openModal } = useModal();
+  const { user } = useWeb3();
+  const search = useSearch({ strict: false }) as any;
```
Comment (tiagofilipenunes) [inline src/components/strategies/create/CreateForm.tsx:73]: Can you set the type of the search here?

**Comment by ashachaf** [comment]
- [ ] (optional) remove "max spread" warning following a price range update
currently, the warning message stays as if the prices were not changed.
updated, once prices are updated, remove the warning message.
https://github.com/bancorprotocol/carbon-app/assets/40430285/332bdf15-0275-4534-a49c-ba7a6ae0dfef

- [ ] text update
current: Buy Low order has been removed
updated: Buy Low order will be remove
`same change for "sell high" warning message`

- [ ] (optional) expose budget change in the edit price flow
add the "step 2: edit budget" element that we have on concentrated liquidity strategy to both disposable and recurring.
few indications:
- it needs to be specific to the budget of each order (i.e. under sell, you can control the sell budget only)
- unlike the concentrated flow where input a value cause recalculation of the 2nd budget, here each input is alone and only effect its own budget input

https://github.com/bancorprotocol/carbon-app/assets/40430285/1632c4cc-baf1-43a5-bb80-5745185c1194

File: src/components/strategies/create/CreateForm.tsx (hunk 1 above)
Comment (GrandSchtroumpf) [reply src/components/strategies/create/CreateForm.tsx:73]: Not here unfortunately since it's used on multiple pages

**Comment by tiagofilipenunes** [comment]
Fantastic PR - a lot was done here. The code is much cleaner, better behaviour overall. Since the goal was to avoid any differences with the main app, I've flagged all the ones I've found below. I'm still finishing code review

- [ ] ErrorWrapper is occupying all vertical height for me, because of changes to `main` `className` in App.tsx
image
image
- [ ] Not used anymore
  - src/components/strategies/create/BuySellBlock/LimitRangeSection.tsx
  - src/components/common/OutsideMarketPriceWarning.tsx
- [ ] Orders overlap boundary condition on recurring creation/edit has changed. I think this way is better, but flagging just in case
  - Before: sell 1500-2000, buy 1000-1500 does not trigger warning, but buy 1000-1501 does
  - Now: sell 1500-2000, buy 1000-1500 triggers a warning
- [ ] Orders reversed boundary condition on recurring creation has changed too. 
  - Before: sell 1000-2000, buy 1000-2000 triggers error
  - After: sell 1000-2000, buy 1000-2000 does not trigger error, but buy 1001-2000 does
- [ ] Buy and sell orders overlap border now shows for all boxes - much better
- [ ] Error top border increased 
Before:
image
After:
image

- [ ] Layout changes on mobile. Now the form spans the whole width
Before:
image
After:
image

- [ ] Bottom border in create form has been reduced
Before:
image
After:
image

- [ ] Smaller space on mobile for chart
Before:
image
After:
image

- [ ] Overlapping create/edit on mobile occupying more than the viewport width
image

- Not a regression, same behaviour as app but just flagging that the overlapping tooltip icons are smaller than disposable and recurring (14x14 vs 18x18)
- [ ] Setting 0 spread in overlapping closes the anchor. 
  - Before: set 0 spread and error is shown, anchor is kept as previous selection
  - After: set 0 spread and error is shown, anchor selection is removed
- [ ] Changing the range to be below or above market price behaviour changed
  - Before: Change the anchor to the other token if the current selection is not possible
  - After: Anchor is reset once the range goes above or below market price, even if the other token anchor selection is possible
- [ ] When price is not available on overlapping strategy, the create strategy button is not shown anymore - I much prefer this. Also, the duplicate warning checkbox no longer shows up.
Before: image
After: 
image

- Current market price selection on overlapping create is kept on refresh - nice!!!
- [ ] Top border on /strategies/create reduced
  - Before: 
image
  - After: 
image
- [ ] Create strategy page form does not have full width on mobile
	- Before: image
	- After: image
- [ ] On edit prices, token logo is shown next to "Sell High"/"Buy Low". I prefer it, looks good
- [ ] Edit/deposit/withdraw overlapping - stays in loading with ";" printed on the screen when api is not available
	- image
- Allocated bar previously only used in overlapping now showing for recurring and disposable edit - much nicer
- [ ] If there is no change to prices on recurring/disposable edit, the button is now disabled with no warning. 
	- Before: user could press create strategy if prices were the same
	- After: CTA is disabled and text shows "Create Strategy"
- [ ] On create strategy mobile, pressing any strategy type option changes the width of the form
	- Before clicking: image
	- After clicking: image

- [ ] On overlapping deposit/withdraw, CTA text changed from "Confirm Deposit"/"Confirm Withdraw" to "Confirm Changes" for both

File: src/components/strategies/edit/EditStrategyBudgetContent.tsx [hunk 2]
```diff
@@ -1,277 +0,0 @@
...
-      onWithdraw: depositOrWithdrawFunds,
-      onDelete: () =>
-        deleteStrategy(strategy, setIsProcessing, () =>
-          carbonEvents.strategyEdit.strategyDelete({
-            strategyId: strategy.id,
-            ...strategyEventData,
-          })
-        ),
-    });
-  };
```
Comment (tiagofilipenunes) [inline src/components/strategies/edit/EditStrategyBudgetContent.tsx:166]: This flow is missing from strategy budget edit. If withdrawing everything -> prompt user to instead delete the strategy

File: src/pages/strategies/create/recurring.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+};
+
+const getError = (search: CreateRecurringStrategySearch) => {
+  const { buyMin, buyMax, sellMin, sellMax } = search;
+  const minReversed = buyMin && sellMin && +buyMin > +sellMin;
+  const maxReversed = buyMax && sellMax && +buyMax > +sellMax;
+  if (minReversed || maxReversed) {
+    return 'Orders are reversed. This strategy is currently set to Buy High and Sell Low. Please adjust your prices to avoid an immediate loss of funds upon creation.';
+  }
+};
```
Comment (tiagofilipenunes) [inline src/pages/strategies/create/recurring.tsx:48]: Can you use checkIfOrdersReversed from src/components/strategies/utils.ts here instead? To keep the check consistent throughout the app

File: src/pages/strategies/create/recurring.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+const getWarning = (search: CreateRecurringStrategySearch) => {
+  const { buyMin, buyMax, sellMin, sellMax } = search;
+  const sellMinInRange =
+    buyMin && buyMax && sellMin && +sellMin >= +buyMin && +sellMin < +buyMax;
+  const buyMaxInRange =
+    sellMin && sellMax && buyMax && +buyMax >= +sellMin && +buyMax < +sellMax;
+  if (sellMinInRange || buyMaxInRange) {
+    return 'Notice: your Buy and Sell orders overlap';
+  }
+};
```
Comment (tiagofilipenunes) [inline src/pages/strategies/create/recurring.tsx:39]: Can you use checkIfOrdersOverlap from src/components/strategies/utils.ts here?

**Review by tiagofilipenunes — CHANGES_REQUESTED** [review]

File: src/components/strategies/edit/EditBudgetFields.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import { useGetTokenBalance } from 'libs/queries';
+import { SafeDecimal } from 'libs/safedecimal';
+import { isZero } from '../common/utils';
+
+interface Props {
+  budget: string;
+  order: BaseOrder;
+  editType: 'withdraw' | 'deposit';
+  buy?: boolean;
+  optional?: boolean;
```
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply src/components/strategies/edit/EditBudgetFields.tsx:21]: It's used here : https://github.com/bancorprotocol/carbon-app/pull/1253/files/eb4264b40d62fcd04418b965583588707161ea63#diff-ea65a52207033fd9c86d08b8363908010d197ae4147eee6db8e52b8f22f77ed1R72

**Review by tiagofilipenunes — APPROVED** [review]

**Comment by ashachaf** [comment]
- [ ] limit order overview grid display
current:
image

updated:
as outlined in this issue: https://github.com/bancorprotocol/carbon-app/issues/773
```Limit order
-- Given a disposable ETH/USDC strategy with buy limit order price of 1500
-- Median price point=buy order price=1500
-- Bucket size=1500/100=15
-- Left border value=1500+15x3=1545
-- Right border value=1500-15x3= 1455
```

- [ ] replace "overlapping" with "concentrated"

- [ ] replace "disposable" with "limit"

**Review by ashachaf — APPROVED** [review]

**Pull request merged**